Assets/Scripts/Algorithms/Patrolling/CognitiveCoordinated/CognitiveCoordinatedVirtualStigmergy.cs

```diff
@@ -0,0 +1,60 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using Maes.Map;
+
+namespace Maes.Algorithms.Patrolling
+{
+    /// <summary>
+    /// This is a distributed version of the Cognitive Coordinated Algorithm.
+    /// It uses virtual stigmergy to coordinate the robots.
+    /// Each robot has its own local map and uses the occupied vertices of other robots to coordinate.
+    /// </summary>
+    public sealed class CognitiveCoordinatedVirtualStigmergy : CognitiveCoordinatedBase
+    {
+        public override string AlgorithmName => "Cognitive Coordinated (virtual-stigmergy knowledge) Algorithm";
+        protected override PatrollingMap _globalMap => _patrollingMap;
+
+        public override void InitializeCoordinator(PatrollingMap _)
+        {
+        }
+
+        public override void OccupyVertex(int robotId, Vertex vertex)
+        {
+#if DEBUG
+            if (!_globalMap.Vertices.Contains(vertex))
+            {
+                throw new ArgumentException($"Vertex ({vertex}) is not a part of GlobalMap.Vertices.", nameof(vertex));
+            }
+#endif
+            if (!_virtualStigmergyComponent.Has(VirtualStigmergyMessageId))
+            {
+                _virtualStigmergyComponent.Put(VirtualStigmergyMessageId, new Dictionary<int, Vertex>());
```

Comment:
Watch out with dictionaries. They are reference types, and this means that when you modify a dictionary, all other virtual stigmergies will have that information, even if they are out of range.

Reply:
Right, it's fixed now, thx